osu.Game/Online/Chat/DrawableLinkCompiler.cs

```diff
@@ -15,30 +19,109 @@ namespace osu.Game.Online.Chat
+            this.link = link;
         }
 
-        [BackgroundDependencyLoader]
-        private void load(OsuColour colours)
+        [BackgroundDependencyLoader(true)]
+        private void load(OsuColour colours, OsuGame game, BeatmapManager beatmapManager)
         {
             IdleColour = colours.Blue;
+            this.game = game;
+            this.beatmapManager = beatmapManager;
         }
 
         protected override IEnumerable<Drawable> EffectTargets => Parts;
 
+        private BeatmapInfo getBeatmapFromLink(LinkDetails link)
+        {
+            int beatmapId = getBeatmapIdFromLink(link);
+            if (beatmapId != 0)
+                return beatmapManager.QueryBeatmap(b => b.OnlineBeatmapID == beatmapId);
+
+            return null;
+        }
+
+        private BeatmapSetInfo getBeatmapSetFromLink(LinkDetails link)
+        {
+            int setId = getBeatmapSetIdFromLink(link);
+            if (setId != 0)
+                return beatmapManager.QueryBeatmap(b => b.BeatmapSet.OnlineBeatmapSetID == setId)?.BeatmapSet;
+
+            return null;
+        }
+
+        private int getBeatmapIdFromLink(LinkDetails link)
+        {
+            if (int.TryParse(link.Argument.Contains('?') ? link.Argument.Split('?')[0] : link.Argument, out int beatmapId))
+                return beatmapId;
+
+            return 0;
+        }
+
+        private int getBeatmapSetIdFromLink(LinkDetails link)
+        {
+            if (int.TryParse(link.Argument, out int setId))
+                return setId;
+
+            return 0;
+        }
+
+        public MenuItem[] ContextMenuItems
+        {
+            get
+            {
+                List<MenuItem> items = new List<MenuItem>();
+
+                switch (link.Action)
+                {
+                    case LinkAction.OpenBeatmap:
+                        BeatmapInfo map = getBeatmapFromLink(link);
+
+                        if (map != null)
+                            items.Add(new OsuMenuItem("Go to beatmap", MenuItemType.Highlighted, () => game?.PresentBeatmap(map.BeatmapSet, b => b.OnlineBeatmapID == map.OnlineBeatmapID)));
+
+                        items.Add(new OsuMenuItem("Details", MenuItemType.Standard, () => game?.ShowBeatmap(getBeatmapIdFromLink(link))));
+
+                        return items.ToArray();
+
+                    case LinkAction.OpenBeatmapSet:
+                        BeatmapSetInfo mapSet = getBeatmapSetFromLink(link);
+
+                        if (mapSet != null)
+                            items.Add(new OsuMenuItem("Go to beatmap set", MenuItemType.Highlighted, () => game?.PresentBeatmap(mapSet)));
+
+                        items.Add(new OsuMenuItem("Details", MenuItemType.Standard, () => game?.ShowBeatmapSet(getBeatmapSetIdFromLink(link))));
+
+                        return items.ToArray();
+
+                    default:
+                        return Array.Empty<MenuItem>();
+                }
+            }
+        }
```

Comment:
And this is problem number 2 in this for me. Going forward I presume we might want to handle news links, wiki links, etc. `DrawableLinkCompiler` is a very general class. I would be strongly opposed to all logic handling dispatch of links to game views landing in this huge getter in the future.

The way I would propose this to be resolved is by having analogues to the link types (so i.e. `DrawableBeatmapLink` and `DrawableBeatmapSetLink`). Each of these will have their own logic for context menu, their own options, etc., all scoped down and not everything-at-once. I don't see this as over-engineering as we know we might need this in the future already.